flax/core/scope.py

```diff
@@ -297,7 +297,15 @@ def variable(self, col: str, name: str, init_fn: Callable[..., T],
                *init_args) -> Variable[T]:
     """Create a Variable."""
     self.reserve(name)
-    if not self.has_variable(col, name):
+    collection = self.collection(col=col, mutable=True)
+
+    # If the collection is immutable and the variable does not exist
+    # in it, we cannot initialize it.
+    if isinstance(collection, FrozenDict) and (name not in collection):
+      raise ValueError(f"Kind: {col} is immutable and "
```

Comment:
Could you use a custom error message instead? https://github.com/google/flax/blob/master/flax/errors.py